torch/_inductor/fx_passes/fuse_attention.py

```diff
@@ -502,6 +502,40 @@ def _sfdp_replacement_18(query, key, value, causal_mask, dropout_p):
     )
 
 
+def _sfdp_pattern_19(query, key, value, causal_mask, attn_mask, dropout_p):
+    # for token-classification+gpt2 / text-generation+gpt2
+    attn_weights = torch.matmul(query, key.permute(0, 1, 3, 2))
+    inv_scale = torch.full(
+        [],
+        value.size(-1) ** 0.5,
+        dtype=attn_weights.dtype,
+        device=attn_weights.device,
+    )
+    attn_weights = attn_weights.div(inv_scale)
+    causal_mask_value = torch.full(
+        (), torch.finfo(query.dtype).min, dtype=query.dtype, device=query.device
+    )
+    attn_weights = torch.where(causal_mask, attn_weights, causal_mask_value)
+    attn_weights = attn_weights + attn_mask
+    attn_weights = attn_weights.softmax(dim=-1).type(value.dtype)
+    return torch.nn.functional.dropout(attn_weights, dropout_p).matmul(value)
+
+
+def _sfdp_replacement_19(query, key, value, causal_mask, attn_mask, dropout_p):
+    counters["inductor"]["fuse_attention"] += 1
+    fill_value = torch.full((), -float("inf"), dtype=query.dtype, device=query.device)
+    attn_mask = torch.where(causal_mask, attn_mask, fill_value)
```

Comment:
And here in the replacement, the causal mask dtype also equals the query dtype. Is the change in attention.cpp really neccessary?

Reply:
Causal mask is bool and attention mask is float. Here is the fx_graph_readable which provides the types. You could see the line of `add_2`: `attn_weights` changes from BF16 to FP32 due to the addition with attention mask (float).

```
# File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:182 in _attn, code: attn_weights = torch.matmul(query, key.transpose(-1, -2))
        permute_3: "bf16[1, 12, 64, 23]" = torch.ops.aten.permute.default(permute_1, [0, 1, 3, 2])
        expand: "bf16[1, 12, 23, 64]" = torch.ops.aten.expand.default(permute, [1, 12, 23, 64]);  permute = None
        view_5: "bf16[12, 23, 64]" = torch.ops.aten.reshape.default(expand, [12, 23, 64]);  expand = None
        expand_1: "bf16[1, 12, 64, 23]" = torch.ops.aten.expand.default(permute_3, [1, 12, 64, 23]);  permute_3 = None
        view_6: "bf16[12, 64, 23]" = torch.ops.aten.reshape.default(expand_1, [12, 64, 23]);  expand_1 = None
        bmm: "bf16[12, 23, 23]" = torch.ops.aten.bmm.default(view_5, view_6);  view_5 = view_6 = None
        view_7: "bf16[1, 12, 23, 23]" = torch.ops.aten.reshape.default(bmm, [1, 12, 23, 23]);  bmm = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:185 in _attn, code: attn_weights = attn_weights / torch.full(
        full_default: "bf16[]" = torch.ops.aten.full.default([], 8.0, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
        div: "bf16[1, 12, 23, 23]" = torch.ops.aten.div.Tensor(view_7, full_default);  view_7 = full_default = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:196 in _attn, code: causal_mask = self.bias[:, :, key_length - query_length : key_length, :key_length].to(torch.bool)
        _frozen_param15: "b8[1, 1, 23, 23]" = self._frozen_param15
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:200 in _attn, code: mask_value = torch.full([], mask_value, dtype=attn_weights.dtype).to(attn_weights.device)
        full_default_1: "bf16[]" = torch.ops.aten.full.default([], -3.3895313892515355e+38, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:201 in _attn, code: attn_weights = torch.where(causal_mask, attn_weights.to(attn_weights.dtype), mask_value)
        where: "bf16[1, 12, 23, 23]" = torch.ops.aten.where.self(_frozen_param15, div, full_default_1);  _frozen_param15 = div = full_default_1 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:205 in _attn, code: attn_weights = attn_weights + attention_mask
        add_2: "f32[1, 12, 23, 23]" = torch.ops.aten.add.Tensor(where, arg14_1);  where = arg14_1 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:207 in _attn, code: attn_weights = nn.functional.softmax(attn_weights, dim=-1)
        amax: "f32[1, 12, 23, 1]" = torch.ops.aten.amax.default(add_2, [-1], True)
        sub_1: "f32[1, 12, 23, 23]" = torch.ops.aten.sub.Tensor(add_2, amax);  add_2 = amax = None
        exp: "f32[1, 12, 23, 23]" = torch.ops.aten.exp.default(sub_1);  sub_1 = None
        sum_1: "f32[1, 12, 23, 1]" = torch.ops.aten.sum.dim_IntList(exp, [-1], True)
        div_1: "f32[1, 12, 23, 23]" = torch.ops.aten.div.Tensor(exp, sum_1);  exp = sum_1 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:210 in _attn, code: attn_weights = attn_weights.type(value.dtype)
        convert_element_type_9: "bf16[1, 12, 23, 23]" = torch.ops.prims.convert_element_type.default(div_1, torch.bfloat16);  div_1 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:211 in _attn, code: attn_weights = self.attn_dropout(attn_weights)
        clone: "bf16[1, 12, 23, 23]" = torch.ops.aten.clone.default(convert_element_type_9);  convert_element_type_9 = None
        
        # File: /home/liaoxuan/transformers/src/transformers/models/gpt2/modeling_gpt2.py:217 in _attn, code: attn_output = torch.matmul(attn_weights, value)
        expand_2: "bf16[1, 12, 23, 23]" = torch.ops.aten.expand.default(clone, [1, 12, 23, 23]);  clone = None
        view_8: "bf16[12, 23, 23]" = torch.ops.aten.reshape.default(expand_2, [12, 23, 23]);  expand_2 = None
        expand_3: "bf16[1, 12, 23, 64]" = torch.ops.aten.expand.default(permute_2, [1, 12, 23, 64])
        view_9: "bf16[12, 23, 64]" = torch.ops.aten.reshape.default(expand_3, [12, 23, 64]);  expand_3 = None
        bmm_1: "bf16[12, 23, 64]" = torch.ops.aten.bmm.default(view_8, view_9);  view_8 = view_9 = None
        view_10: "bf16[1, 12, 23, 64]" = torch.ops.aten.reshape.default(bmm_1, [1, 12, 23, 64]);  bmm_1 = None
```